### PR TITLE
feat: enable output of unmapped reads

### PR DESCRIPTION
**Pull request by kashyapchhatbar**
### Description

Ability to save unmapped reads in fastq or fastq.gz formats using the `--outReadsUnmapped Fastx` option in STAR

### QC

* [x] I confirm that:

For all wrappers added by this PR, 

* there is a test case which covers any introduced changes,
* `input:` and `output:` file paths in the resulting rule can be changed arbitrarily,
* either the wrapper can only use a single core, or the example rule contains a `threads: x` statement with `x` being a reasonable default,
* rule names in the test case are in [snake_case](https://en.wikipedia.org/wiki/Snake_case) and somehow tell what the rule is about or match the tools purpose or name (e.g., `map_reads` for a step that maps reads),
* all `environment.yaml` specifications follow [the respective best practices](https://stackoverflow.com/a/64594513/2352071),
* wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`),
* all fields of the example rules in the `Snakefile`s and their entries are explained via comments (`input:`/`output:`/`params:` etc.),
* `stderr` and/or `stdout` are logged correctly (`log:`), depending on the wrapped tool,
* temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to (see [here](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir); this also means that using any Python `tempfile` default behavior works),
* the `meta.yaml` contains a link to the documentation of the respective tool or command,
* `Snakefile`s pass the linting (`snakemake --lint`),
* `Snakefile`s are formatted with [snakefmt](https://github.com/snakemake/snakefmt),
* Python wrapper scripts are formatted with [black](https://black.readthedocs.io).
* Conda environments use a minimal amount of channels, in recommended ordering. E.g. for bioconda, use (conda-forge, bioconda, nodefaults, as conda-forge should have highest priority and defaults channels are usually not needed because most packages are in conda-forge nowadays).

### CODE REVIEW

File: bio/star/align/wrapper.py [hunk 1]
```diff
@@ -42,6 +42,12 @@
 else:
     readcmd = ""
 
+if snakemake.output.get("unmapped1") and snakemake.output.get("unmapped2"):
+    unfq1 = snakemake.output.get("unmapped1")
+    unfq2 = snakemake.output.get("unmapped2")
```
Comment (fgvieira) [inline bio/star/align/wrapper.py:47]: These do not seem to be used anywhere.
```suggestion
```

**Review by kashyapchhatbar — COMMENTED** [review]
Thanks for making it clearer.

**Comment by kashyapchhatbar** [comment]
I have updated the files and all the tests are running and making sense. Hopefully, it is good to merge now.

File: bio/star/align/wrapper.py [hunk 2]
```diff
@@ -42,6 +42,12 @@
 else:
     readcmd = ""
 
+if snakemake.output.get("unmapped1") and snakemake.output.get("unmapped2"):
+    unmappedcmd = "--outReadsUnmapped Fastx"
+elif snakemake.output.get("unmapped"):
+    unmappedcmd = "--outReadsUnmapped Fastx"
+else:
+    unmappedcmd = ""
```
Comment (fgvieira) [inline bio/star/align/wrapper.py:50]:
```suggestion
out_unmapped = snakemake.output.get("unmapped", "")
if out_unmapped:
    out_unmapped = "--outReadsUnmapped Fastx"
```

File: bio/star/align/wrapper.py [hunk 3]
```diff
@@ -64,6 +70,7 @@
         " --readFilesIn {input_str}"
         " {readcmd}"
         " {extra}"
+        " {unmappedcmd}"
```
Comment (fgvieira) [inline bio/star/align/wrapper.py:73]:
```suggestion
        " {out_unmapped}"
```

File: bio/star/align/wrapper.py [hunk 4]
```diff
@@ -83,3 +90,22 @@
...
+            shell(
+                "gzip -c {tmpdir}/Unmapped.out.mate2 > {snakemake.output.unmapped2:q}"
+            )
+        else:
+            shell("cat {tmpdir}/Unmapped.out.mate2 > {snakemake.output.unmapped2:q}")
+    if snakemake.output.get("unmapped"):
+        if snakemake.output.get("unmapped").endswith("gz"):
+            shell("gzip -c {tmpdir}/Unmapped.out.mate1 > {snakemake.output.unmapped:q}")
+        else:
+            shell("cat {tmpdir}/Unmapped.out.mate1 > {snakemake.output.unmapped:q}")
```
Comment (fgvieira) [inline bio/star/align/wrapper.py:111]:
```suggestion
    unmapped = snakemake.output.get("unmapped")
    if unmapped:
        # SE
        if not fq2:
            unmapped = [unmapped]

    for i, out_unmapped in unmapped.items():
        if out_unmapped.endswith("gz"):
            shell(
                "gzip -c {tmpdir}/Unmapped.out.mate{i+1} > {snakemake.output.unmapped[i]:q}"
            )
        else:
            shell("cat {tmpdir}/Unmapped.out.mate{i+1} > {snakemake.output.unmapped[i]:q}")
```

File: bio/star/align/test/Snakefile [hunk 5]
```diff
@@ -12,6 +12,8 @@ rule star_pe_multi:
         aln="star/pe/{sample}/pe_aligned.sam",
         log="logs/pe/{sample}/Log.out",
         sj="star/pe/{sample}/SJ.out.tab",
+        unmapped1="star/pe/{sample}/unmapped.1.fastq.gz",
+        unmapped2="star/pe/{sample}/unmapped.2.fastq.gz",
```
[the remark this answers is not in the document]
Comment (fgvieira) [reply bio/star/align/test/Snakefile:16]: I think it would be nice that both for SE and PE the input files stay the same.

**Comment by kashyapchhatbar** [comment]
While it is a nice attempt to convert the unmapped files to a list, it doesn't work out nicely downstream. The test fails because the shell function from snakemake doesn't deal with ints properly.

File: bio/star/align/wrapper.py [hunk 6]
```diff
@@ -83,3 +87,18 @@
         shell("cat {tmpdir}/Log.progress.out > {snakemake.output.log_progress:q}")
     if snakemake.output.get("log_final"):
         shell("cat {tmpdir}/Log.final.out > {snakemake.output.log_final:q}")
+    unmapped = snakemake.output.get("unmapped")
+    if unmapped:
+        # SE
+        if not fq2:
+            unmapped = [unmapped]
+
+    for i, out_unmapped in unmapped.items():
```
Comment (fgvieira) [inline bio/star/align/wrapper.py:96]:
```suggestion
    for i, out_unmapped in enumerate(unmapped):
```

**Comment by kashyapchhatbar** [comment]
This will still not work because `shell` function converts everything to string and that's why you can't do the `i+1` thing

File: bio/star/align/wrapper.py [hunk 7]
```diff
@@ -83,3 +87,18 @@
...
+            unmapped = [unmapped]
+
+    for i, out_unmapped in enumerate(unmapped):
+        if out_unmapped.endswith("gz"):
+            shell(
+                "gzip -c {tmpdir}/Unmapped.out.mate{i+1} > {snakemake.output.unmapped[i]:q}"
+            )
+        else:
+            shell(
+                "cat {tmpdir}/Unmapped.out.mate{i+1} > {snakemake.output.unmapped[i]:q}"
```
Comment (fgvieira) [inline bio/star/align/wrapper.py:103]:
```suggestion
        j = i + 1
        cmd = "gzip -c" if out_unmapped.endswith("gz") else "cat":
        shell(
            "{cmd} {tmpdir}/Unmapped.out.mate{j} > {snakemake.output.unmapped[i]:q}"
        )
```

File: bio/star/align/wrapper.py [hunk 8]
```diff
@@ -83,3 +87,18 @@
...
+            )
```
Comment (fgvieira) [inline bio/star/align/wrapper.py:104]:
```suggestion
    for i, out_unmapped in enumerate(unmapped, 1):
        cmd = "gzip -c" if out_unmapped.endswith("gz") else "cat":
        shell(
            "{cmd} {tmpdir}/Unmapped.out.mate{i} > {snakemake.output.unmapped[i-1]:q}"
        )
```

File: bio/star/align/wrapper.py [hunk 9]
```diff
@@ -83,3 +87,14 @@
...
     if snakemake.output.get("log_final"):
         shell("cat {tmpdir}/Log.final.out > {snakemake.output.log_final:q}")
+    unmapped = snakemake.output.get("unmapped")
+    if unmapped:
+        # SE
+        if not fq2:
+            unmapped = [unmapped]
+
+    for i, out_unmapped in enumerate(unmapped, 1):
+        cmd = "gzip -c" if out_unmapped.endswith("gz") else "cat":
```
Comment (fgvieira) [inline bio/star/align/wrapper.py:97]:
```suggestion
        cmd = "gzip -c" if out_unmapped.endswith("gz") else "cat"
```

File: bio/star/align/wrapper.py [hunk 10]
```diff
@@ -83,3 +87,14 @@
...
+    if unmapped:
+        # SE
+        if not fq2:
+            unmapped = [unmapped]
+
+    for i, out_unmapped in enumerate(unmapped, 1):
+        cmd = "gzip -c" if out_unmapped.endswith("gz") else "cat"
+        shell(
+            "{cmd} {tmpdir}/Unmapped.out.mate{i} > {snakemake.output.unmapped[i-1]:q}"
+        )
```
Comment (fgvieira) [inline bio/star/align/wrapper.py:100]:
```suggestion
        for i, out_unmapped in enumerate(unmapped, 1):
            cmd = "gzip -c" if out_unmapped.endswith("gz") else "cat"
            shell(
                "{cmd} {tmpdir}/Unmapped.out.mate{i} > {snakemake.output.unmapped[i-1]:q}"
            )
```

**Comment by kashyapchhatbar** [comment]
Thanks, I won't be able to check with -p command since it'll try to grab the old-version wrapper.py which won't have the out_unmapped command and will fail. I guess since the `test/Snakefile` is updated with unmapped files as output and if they are passing, it should be working.

**Review by fgvieira — APPROVED** [review]

File: bio/star/align/wrapper.py [hunk 11]
```diff
@@ -42,6 +42,9 @@
 else:
     readcmd = ""
 
+out_unmapped = snakemake.output.get("unmapped", None)
```
Comment (fgvieira) [inline bio/star/align/wrapper.py:45]:
```suggestion
out_unmapped = snakemake.output.get("unmapped", "")
```

**Pull request merged**